compiler/front/optionsprocessor.nim

```diff
@@ -0,0 +1,1836 @@
+      # xxx: we hack commandLineSrcIdx at callers like `nimconf` to get different
+      #      info here; rework so it's all handled via returns and remove the
+      #      need for info.
+    try:
+      AbsoluteDir conf.pathSubs(p, conf.toFullPath(info).splitFile().dir)
+    except ValueError:
+      result = ProcSwitchResult(kind: procSwitchErrArgPathInvalid,
+                                switch: result.switch,
+                                givenSwitch: s,
+                                givenArg: arg,
+                                pathAttempted: p,
+                                srcCodeOrigin: instLoc())
+      return
+
+  template argProcessCfgPath(conf: ConfigRef, arg, s: string): AbsoluteDir =
+    let
+      path = if arg.len > 0 and arg[0] == '"': strutils.unescape(arg)
+             else: arg
+      info = newLineInfo(conf.commandLineSrcIdx, 0, -1)
+      # xxx: we hack commandLineSrcIdx at callers like `nimconf` to get different
+      #      info here; rework so it's all handled via returns and remove the
+      #      need for info.
+      basedir = toFullPath(conf, info).splitFile().dir
+      p = if os.isAbsolute(path) or '$' in path:
+              path
+            else:
+              basedir / path
+    try:
+      AbsoluteDir pathSubs(conf, p, basedir)
+    except ValueError:
+      result = ProcSwitchResult(kind: procSwitchErrArgPathInvalid,
+                                switch: result.switch,
+                                givenSwitch: s,
+                                givenArg: arg,
+                                pathAttempted: p,
+                                srcCodeOrigin: instLoc())
+      return
+
+  template argSplit(s, arg: string; key, val: var string) =
+    if not splitSwitch(arg, key, val):
+      result = ProcSwitchResult(kind: procSwitchErrArgMalformedKeyValPair,
+                                switch: result.switch,
+                                givenSwitch: s,
+                                givenArg: arg,
+                                srcCodeOrigin: instLoc())
+      return
+
+  template switchOn(s, arg: string): bool =
+    case arg.normalize
+    of "", "on": true
+    of "off": false
+    else:
+      result = ProcSwitchResult(kind: procSwitchErrArgExpectedOnOrOff,
+                                switch: result.switch,
+                                givenSwitch: s,
+                                givenArg: arg,
+                                srcCodeOrigin: instLoc())
+      return
+
+  template processOnOffSwitch(conf: ConfigRef, op: TOptions, arg, s: string) =
+    case arg.normalize
+    of "", "on": conf.incl op
+    of "off": conf.excl op
+    else:
+      result = ProcSwitchResult(kind: procSwitchErrArgExpectedOnOrOff,
+                                switch: result.switch,
+                                givenSwitch: s,
+                                givenArg: arg,
+                                srcCodeOrigin: instLoc())
+      return
```

Comment:
```suggestion
    {.line.}:
      case switchOn(s, arg)
      of true:  conf.incl op
      of false: conf.excl op
```
Soft suggestion. It slightly reduces code duplication.

Reply:
I like it!